crates/plugins/src/error.rs

```diff
@@ -0,0 +1,89 @@
+pub type PluginLookupResult<T> = std::result::Result<T, Error>;
+
+/// Error message during plugin lookup or deserializing
+#[derive(Debug, thiserror::Error)]
+pub enum Error {
+    #[error("{0}")]
+    NotFound(NotFoundError),
+
+    #[error("{0}")]
+    ConnectionFailed(ConnectionFailedError),
+
+    #[error("{0}")]
+    InvalidManifest(InvalidManifestError),
+
+    #[error("URL parse error {0}")]
+    UrlParseError(url::ParseError),
+}
+
+/// Contains error details for when a plugin resource cannot be found at expected location
+#[derive(Debug)]
+pub struct NotFoundError {
+    name: Option<String>,
+    addr: String,
+    err: String,
+}
+
+impl NotFoundError {
+    pub fn new(name: Option<String>, addr: String, err: String) -> Self {
+        Self { name, addr, err }
+    }
+}
+
+impl std::fmt::Display for NotFoundError {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        f.write_str(&format!(
+            "plugin '{}' not found at expected location {}: {}",
+            self.name.clone().unwrap_or_default(),
```

Comment:
Makes no practical difference in this specific case, but as a general pattern that comes up pretty frequently:

```suggestion
            self.name.as_deref().unwrap_or_default(),
```

`as_deref` borrows an `Option<&str>` from the `Option<String>`.